### ISSUE TITLE
Understanding performance of deflate in libewf (ewfacquirestream)

### ISSUE THREAD

**Issue by JustMyGithub** [libyal/libewf#178]
Hey there,

I just built the current version and I notice that compression speed is pretty slow with best (deflate) compression.
It starts at about 10MBs and after 10GB it is at about 35-40 MBs. It does not seem to matter whether I use -j 1, -j 12 or -j 50 -- ewfacquire uses about 8-10% of 6x2 cores only. The legacy-build achieves the same performance.
I do expect low performance for best compression, of course. However CPU is hardly utilized and so are source and destination drives. If the compression is parallelized (at least for the non-legacy version - why would there be a -j parameter otherwise), what is the bottleneck and can I do something about it using different configuration parameters?

Thx

**Comment by joachimmetz** [comment]
I'm not able to tell, too little information about your configuration, build and the test parameters. The code is considered experimental, you could use normal code profiling tools to see what is going on.

**Comment by JustMyGithub** [comment]
I understand that non-legacy version is experimental. I tried this version because non-legacy may not be parallelized based on my understanding of the documentation, but as outlined above the performance is about the same.

I do not have experience with code profiling for performance aspects, but I am happy to provide the information you need. I assume the most helpful information is the ./configure summary. May it be useful to use libcthreads instead of pthreads?:

```
Building:
   libcerror support:                        local
   libcthreads support:                      local
   libcdata support:                         local
   libcdatetime support:                     local
   libclocale support:                       local
   libcnotify support:                       local
   libcsplit support:                        local
   libuna support:                           local
   libcfile support:                         local
   libcpath support:                         local
   libbfio support:                          local
   libfcache support:                        local
   libfdata support:                         local
   libfdatetime support:                     local
   libfguid support:                         local
   libfvalue support:                        local
   ADLER32 checksum support:                 zlib
   DEFLATE compression support:              zlib
   BZIP2 compression support:                bzlib
   libhmac support:                          local
   MD5 support:                              libcrypto_evp
   SHA1 support:                             libcrypto_evp
   SHA256 support:                           libcrypto_evp
   libcaes support:                          local
   libodraw support:                         local
   libsmdev support:                         local
   libsmraw support:                         local
   GUID/UUID support:                        check
   FUSE support:                             no

Features:
   Multi-threading support:                  pthread
   Wide character type support:              no
   Write support:                            yes
   ewftools are build as static executables: no
   Python (pyewf) support:                   no
   Verbose output:                           no
   Debug output:                             no
```

**Comment by joachimmetz** [comment]
> I tried this version because non-legacy may not be parallelized based on my understanding of the documentation, but as outlined above the performance is about the same.

This should not be the case, but you might be hitting a yet unimplemented use case. Can you describe step by step:

1. how you build the code, including attaching config.log
2. what commands you ran, and the start of their output
3. other details that might be relevant. 

> May it be useful to use libcthreads instead of pthreads?:

Libcthreads is a cross-platform wrapper around different thread implementations it uses the pthread (without s) implementation under the hood on some configurations.

**Comment by JustMyGithub** [comment]
[legacy_config.log](https://github.com/libyal/libewf/files/10406812/legacy_config.log)
[non-legacy_config.log](https://github.com/libyal/libewf/files/10406818/non-legacy_config.log)

1. In cygwin following https://github.com/libyal/libewf/wiki/Building
so basically:
```
./synclibs.sh
./autogen.sh
./configure
make
```
2. 

```
pbzip2.exe -d --stdout /cygdrive/d/testimage.bz2 | ~/libewf-legacy-main/ewftools/ewfacquirestream -c deflate:best -S 25G -l /cygdrive/k/libewf-legacy.log -d sha256 -t /cygdrive/k/image-legacy
```

(pbzip2 is not the bottleneck. If I use compression fast for example, the throughput is much higher; for non-legacy i used the same parameters but added -j for testing as written in previous post.; d and k are different drives and hardly utilized, so not the bottleneck either.)

```
ewfacquirestream 20140815

Using the following acquiry parameters:
Image path and filename:                /cygdrive/k/image-legacy.E01
Case number:                            case_number
Description:                            description
Evidence number:                        evidence_number
Examiner name:                          examiner_name
Notes:                                  notes
Media type:                             fixed disk
Is physical:                            yes
EWF file format:                        EnCase 6 (.E01)
Compression method:                     deflate
Compression level:                      best
Acquiry start offset:                   0
Number of bytes to acquire:             0 (until end of input)
Evidence segment file size:             25 GiB (26843545600 bytes)
Bytes per sector:                       512
Block size:                             64 sectors
Error granularity:                      64 sectors
Retries on read error:                  2
Zero sectors on read error:             no

Acquiry started at: Jan 11, 2023 10:47:33
This could take a while.

Status: acquired 74 MiB (78610432 bytes)
        in 4 second(s) with 18 MiB/s (19652608 bytes/second).

Status: acquired 104 MiB (109379584 bytes)
        in 8 second(s) with 13 MiB/s (13672448 bytes/second).

Status: acquired 134 MiB (140738560 bytes)
        in 12 second(s) with 11 MiB/s (11728213 bytes/second).
[...]
Status: acquired 483 GiB (519397568992 bytes)
        in 3 hour(s), 33 minute(s) and 10 second(s) with 38 MiB/s (40613857 bytes/second).
```

Thanks

**Comment by joachimmetz** [comment]
20140815 is the legacy version

**Comment by JustMyGithub** [comment]
That is correct. As described above, it does not really make a difference in the result (besides the version number) and I did not want to SPAM. Reproducing it however, I did find out that -j 50 defaults to -j 4 -- I did not notice that before. More than 32 jobs default to 4, so the effective maximum I tried was -12. (Wouldn't it be useful to default to max instead of default if the given parameter is > max?)

Here is the start of output for -j 32 with non-legacy
```
ewfacquirestream 20230101

Using the following acquiry parameters:
Image path and filename:                /cygdrive/k/image.E01
Case number:                            case_number
Description:                            description
Evidence number:                        evidence_number
Examiner name:                          examiner_name
Notes:                                  notes
Media type:                             fixed disk
Is physical:                            yes
EWF file format:                        EnCase 6 (.E01)
Compression method:                     deflate
Compression level:                      best
Acquiry start offset:                   0
Number of bytes to acquire:             0 (until end of input)
Evidence segment file size:             25 GiB (26843545600 bytes)
Bytes per sector:                       512
Block size:                             64 sectors
Error granularity:                      64 sectors
Retries on read error:                  2
Zero sectors on read error:             no

Acquiry started at: Jan 13, 2023 07:45:40
This could take a while.

Status: acquired 70 MiB (74022912 bytes)
        in 4 second(s) with 17 MiB/s (18505728 bytes/second)

Status: acquired 95 MiB (100499456 bytes)
        in 8 second(s) with 11 MiB/s (12562432 bytes/second)

Status: acquired 123 MiB (129859584 bytes)
        in 12 second(s) with 10 MiB/s (10821632 bytes/second)
[...]
Status: acquired 12 GiB (13176012800 bytes)
        in 6 minute(s) and 16 second(s) with 33 MiB/s (35042587 bytes/second)


```

**Comment by joachimmetz** [comment]
Would recommend to try with `-x`. Based on some quick tests on my end, where:

* /dev/zero is optimal case
* /dev/urandom is worst case

```
dd if=/dev/urandom of=random.raw bs=$(( 64 * 512 )) count=$(( 52428800 / 64 ))
26843545600 bytes (27 GB, 25 GiB) copied, 57.7454 s, 465 MB/s

ewfacquirestream -j 8 -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/zero
Written: 25 GiB (26843546916 bytes) in 40 second(s) with 640 MiB/s (671088672 bytes/second)
(per perf utility most of the time spent in libcrypto)

ewfacquirestream -j 8 -x -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/zero
Written: 25 GiB (26843546916 bytes) in 36 second(s) with 711 MiB/s (745654081 bytes/second)

ewfacquirestream -j 8 -x -c deflate:empty-block -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/zero
Written: 25 GiB (26843546916 bytes) in 39 second(s) with 656 MiB/s (688296074 bytes/second)

[libewf-legacy with zlib-ng]
ewfacquirestream -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 4 minute(s) and 24 second(s) with 96 MiB/s (101680097 bytes/second).

[buffered read/write and 8 jobs with zlib-ng]
ewfacquirestream -j 8 -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 3 minute(s) and 6 second(s) with 137 MiB/s (144320138 bytes/second)
(per perf utility most of the time spent in deflate_fast and related functions)

[buffered read/write and 16 jobs with zlib-ng]
ewfacquirestream -j 16 -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 3 minute(s) and 10 second(s) with 134 MiB/s (141281819 bytes/second)
(per perf utility most of the time spent in deflate_fast and related functions)

[chunk read/write (-x) and 8 jobs with zlib-ng]
ewfacquirestream -j 8 -x -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 1 minute(s) and 27 second(s) with 294 MiB/s (308546503 bytes/second)

[chunk read/write (-x) and 16 jobs with zlib-ng]
ewfacquirestream -j 16 -x -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 1 minute(s) and 31 second(s) with 281 MiB/s (294984019 bytes/second)

[buffered read/write and 8 jobs with zlib-1.3.2]
ewfacquirestream -j 8 -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 4 minute(s) and 31 second(s) with 94 MiB/s (99053674 bytes/second)
(per perf utility most of the time spent in deflate_fast and related functions)

[buffered read/write and 8 jobs with zlib-ng]
ewfacquirestream -j 8 -c deflate:best -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 4 minute(s) and 29 second(s) with 95 MiB/s (99790133 bytes/second)
(per perf utility most of the time spent in deflate_best and related functions)

[chunk read/write (-x) and 8 jobs with zlib-ng]
ewfacquirestream -j 8 -x -c deflate:best -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 1 minute(s) and 28 second(s) with 290 MiB/s (305040293 bytes/second)
```

**Comment by JustMyGithub** [comment]
Thanks, that looks promising, thank you. It seems like -x is not in the docs/man page, what does it do?

And did you consider the suggestion "Wouldn't it be useful to default to max instead of default if the given parameter is > max?"

**Comment by joachimmetz** [comment]
> Thanks, that looks promising, thank you. It seems like -x is not in the docs/man page, what does it do?

??? https://github.com/libyal/libewf/blob/main/manuals/ewfacquire.1#L136 or https://github.com/libyal/libewf/blob/main/manuals/ewfacquirestream.1#L112

> And did you consider the suggestion "Wouldn't it be useful to default to max instead of default if the given parameter is > max?"

What would be a sensible max? 

```
[chunk read/write (-x) and 8 jobs with zlib-ng]
ewfacquirestream -j 8 -x -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 1 minute(s) and 27 second(s) with 294 MiB/s (308546503 bytes/second)

[chunk read/write (-x) and 16 jobs with zlib-ng]
ewfacquirestream -j 16 -x -c deflate:fast -S 25G -B $(( 25 * 1024 * 1024 * 1024 ))  -d sha256 -t test < /dev/urandom
Written: 25 GiB (26843545788 bytes) in 1 minute(s) and 31 second(s) with 281 MiB/s (294984019 bytes/second)
```

Indicates/hints more is not necessarily better

**Comment by JustMyGithub** [comment]
Thanks, so for other future readers:
>use the data chunk functions instead of the buffered read and write functions.

so it is unbuffered I/O.

I think my wording was misleading. Let me rephrase: It is not intuitive, that -j 50 sets -j 4 but -j 32 is possible. If I set a value too large I'd expect that the largest possible value is used or the software terminates with an error. 
So if -j 32 is the limit (because it does not scale well beyond that point, because that is the number of virtual cores, what ever) and I set -j 50, I would expect a correction to -j 32 and not -j 4. And I guess that even though in your test -j 16 is slightly slower than -j 8, it is still faster than -j 4, isn't it`t?
That was my suggestion.

**Comment by joachimmetz** [comment]
> I think my wording was misleading. 

Thanks for the clarification. I cannot look into your thoughts to parse what you are thinking so please be explicit in your communication, especially asynchronous one.

Given that this project is very much experimental I'll take a look at a later date at potential UX improvements. Also see https://github.com/libyal/libewf/issues/42.
